pkg/util/winutil/service.go

```diff
@@ -0,0 +1,241 @@
+// Copyright 2016-present Datadog, Inc.
+
+//go:build windows
+// +build windows
+
+package winutil
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"time"
+	"unsafe"
+
+	"golang.org/x/sys/windows"
+	"golang.org/x/sys/windows/svc"
+	"golang.org/x/sys/windows/svc/mgr"
+
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+const (
+	defaultServiceCommandTimeout = 10
+)
+
+var (
+	modadvapi32               = windows.NewLazyDLL("advapi32.dll")
+	procEnumDependentServices = modadvapi32.NewProc("EnumDependentServicesW")
+)
+
+type enumServiceState uint32
+
+func OpenSCManager(desiredAccess uint32) (*mgr.Mgr, error) {
+	h, err := windows.OpenSCManager(nil, nil, desiredAccess)
+	if h == 0 {
+		return nil, err
+	}
+	return &mgr.Mgr{Handle: h}, err
+}
+
+func OpenService(manager *mgr.Mgr, serviceName string, desiredAccess uint32) (*mgr.Service, error) {
+	h, err := windows.OpenService(manager.Handle, windows.StringToUTF16Ptr(serviceName), desiredAccess)
+	if err != nil {
+		return nil, err
+	}
+	return &mgr.Service{Name: serviceName, Handle: h}, err
+}
+
+// Start serviceName via SCM
+// Does not block until service is started
+// https://learn.microsoft.com/en-us/windows/win32/api/winsvc/nf-winsvc-startservicea#remarks
+func StartService(serviceName string) error {
+
+	m, err := OpenSCManager(windows.SC_MANAGER_CONNECT)
+	if err != nil {
+		return err
+	}
+	defer m.Disconnect()
+
+	s, err := OpenService(m, serviceName, windows.SERVICE_START)
+	if err != nil {
+		return fmt.Errorf("could not open service: %v", err)
+	}
+	defer s.Close()
+
+	err = s.Start()
+	if err != nil {
+		return fmt.Errorf("could not start service: %v", err)
+	}
+	return nil
+}
+
+// This function sends a control code to a specified service and waits up to
+// timeout for the service to transition to the requested state
+func ControlService(serviceName string, command svc.Cmd, to svc.State, desiredAccess uint32, timeout int64) error {
+
+	m, err := OpenSCManager(windows.SC_MANAGER_CONNECT)
+	if err != nil {
+		return err
+	}
+	defer m.Disconnect()
+
+	s, err := OpenService(m, serviceName, desiredAccess)
+	if err != nil {
+		return fmt.Errorf("could not open service: %v", err)
+	}
+	defer s.Close()
+
+	status, err := s.Control(command)
+	if err != nil {
+		return fmt.Errorf("could not send control=%d: %v", svc.Stop, err)
+	}
+
+	timesup := time.Now().Add(time.Duration(timeout) * time.Second)
+	for status.State != to {
+		if time.Now().After(timesup) {
+			return fmt.Errorf("timeout waiting for service to go to state=%d", to)
+		}
+		time.Sleep(300 * time.Millisecond)
+		status, err = s.Query()
+		if err != nil {
+			return fmt.Errorf("could not retrieve service status: %v", err)
+		}
+	}
+	return nil
+}
+
+func doStopService(serviceName string) error {
+	return ControlService(serviceName, svc.Stop, svc.Stopped, windows.SERVICE_STOP|windows.SERVICE_QUERY_STATUS, defaultServiceCommandTimeout)
+}
+
+func StopService(serviceName string) error {
+
```

Comment:
I wonder if you can create a single service array list ordered from non-depended to more depended services. I am not 100% sure but don't you need to collect dependents of the dependents (and order them appropriately)?

Reply:
> I wonder if you can create a single service array list ordered from non-depended to more depended services.

That is what `ListDependentServices` returns for us. It wraps `enumDependentServices`, which returns a slice of all dependent services in "reverse start order". From what I can tell, no, we don't need to collect dependents of dependents - they are all returned at once.

[Docs](https://learn.microsoft.com/en-us/windows/win32/api/winsvc/nf-winsvc-enumdependentservicesa)
[Example code](https://learn.microsoft.com/en-us/windows/win32/services/stopping-a-service)
[Unit tests](https://github.com/DataDog/datadog-agent/blob/richl/windows-service-utils/pkg/util/winutil/service_test.go#L91)

Reply:
Interesting. I did not realize it. The very minor suggestion from previous is still possible but it is a nitpicking. All services can be stuffed into array and looped the same way. Would save few lines at least.

Reply:
Ah, I see what you're saying. I think I'd prefer to keep it as-is, so that its clear to the reader that first all the dependent services are being stopped, and then the desired service is stopped.